aleph/logic/export.py

```diff
@@ -0,0 +1,89 @@
+import io
+import logging
+import os
+import mimetypes
+
+import requests
+import zipstream
+
+from followthemoney.export.csv import (
+    write_entity as write_entity_csv, write_headers
+)
+from followthemoney.export.excel import (
+    get_workbook, get_sheet, write_entity as write_entity_excel,
+    get_workbook_content,
+)
+
+
+FORMAT_CSV = 'csv'
+FORMAT_EXCEL = 'excel'
+
+
+log = logging.getLogger(__name__)
+
+
+def write_document(zip_archive, entity):
+    parent = entity.context.get('collection')['label']
+    name = entity.context.get('name')
+    filetypes = entity.context.get('mimetypes')
+    # is it a folder?
+    if 'inode/directory' in filetypes:
+        return
+    ext = mimetypes.guess_extension(filetypes[0], strict=True)
+    if ext:
+        name = name + ext
+    path = os.path.join(parent, name)
+
+    file_url = entity.context['links'].get('file')
+    if file_url:
+        stream = requests.get(file_url, stream=True)
+        zip_archive.write_iter(path, stream.iter_content())
+
+
+def export_entity_csv(handlers, entity):
+    fh = handlers.get(entity.schema.plural)
+    if fh is None:
+        handlers[entity.schema.plural] = fh = io.StringIO()
+        write_headers(fh, entity.schema, extra_headers=['url'])
+    if 'file' in entity.context['links']:
+        url = entity.context['links']['file']
+    else:
+        url = entity.context['links']['ui']
+    write_entity_csv(fh, entity, extra_fields={'url': url})
+
+
+def export_entity_excel(workbook, entity):
+    sheet = get_sheet(entity.schema, workbook, extra_headers=['url'])
+    if 'file' in entity.context['links']:
+        url = entity.context['links']['file']
+    else:
+        url = entity.context['links']['ui']
+    write_entity_excel(sheet, entity, extra_fields={'url': url})
+
+
+def export_entities(entities, format):
+    assert format in (FORMAT_CSV, FORMAT_EXCEL)
+    zip_archive = zipstream.ZipFile()
+
+    if format == FORMAT_EXCEL:
+        workbook = get_workbook()
+        for entity in entities:
+            export_entity_excel(workbook, entity)
+            if entity.schema.is_a('Document'):
+                write_document(zip_archive, entity)
```

Comment:
Looking at the code for ZipStream, I feel we could basically unwind here and make it so that each document is yielded individually, so that we never keep more than one document in memory. Check this loop, which we could essentially plunder (and then emit the metadata last):

see https://github.com/SpiderOak/ZipStream/blob/master/zipstream.py#L200

Reply:
We are using this: https://github.com/allanlei/python-zipstream/blob/master/zipstream/__init__.py#L182. Yielding data from individual documents one at a time is already implemented internally.